src/BTVNanoCommissioning/workflows/BTA_ttbar_producer.py

```diff
@@ -0,0 +1,527 @@
+                "m": ak.concatenate([eles_pass.mass, muons_pass.mass], axis=1),
+                "ch": ak.concatenate([eles_pass.charge, muons_pass.charge], axis=1),
+                "id": ak.concatenate(
+                    [abs(eles_pass.pdgId), abs(muons_pass.pdgId)], axis=1
+                ),
+            }
+        )
+        if not isRealData:
+            lep_arrays["gid"] = ak.fill_none(
+                ak.concatenate(
+                    [eles_pass.matched_gen.pdgId, muons_pass.matched_gen.pdgId],
+                    axis=1,
+                ),
+                0,
+            )
+
+        if not isRealData:
+            ############
+            #  Genlep  #
+            ############
+            # Genlep: same with nominal BTA workflow
+            _fix = lambda x: ak.fill_none(x, 0)
+            is_lep = (
+                lambda p: (abs(_fix(p.pdgId)) == 11)
+                | (abs(_fix(p.pdgId)) == 13)
+                | (abs(_fix(p.pdgId)) == 15)
+            )
+            is_WZ = lambda p: (abs(_fix(p.pdgId)) == 23) | (abs(_fix(p.pdgId)) == 24)
+            is_heavy_hadron = lambda p, pid: (abs(_fix(p.pdgId)) // 100 == pid) | (
+                abs(_fix(p.pdgId)) // 1000 == pid
+            )
+
+            sel = (
+                is_lep(events.GenPart)
+                & ((events.GenPart.statusFlags & (1 << 13)) > 0)
+                & (events.GenPart.pt > 3.0)
+            )  # requires pT > 3 GeV
+            genlep = events.GenPart[sel]
+
+            # trace parents up to 4 generations (from BTA code)
+            genlep_pa1G = genlep.parent
+            genlep_pa2G = genlep.parent.parent
+            genlep_pa3G = genlep.parent.parent.parent
+            genlep_pa4G = genlep.parent.parent.parent.parent
+            istau = genlep_pa1G.pdgId == 15
+            isWZ = is_WZ(genlep_pa1G) | is_WZ(genlep_pa2G)
+            isD = is_heavy_hadron(genlep_pa1G, 4) | is_heavy_hadron(genlep_pa2G, 4)
+            isB = (
+                is_heavy_hadron(genlep_pa1G, 5)
+                | is_heavy_hadron(genlep_pa2G, 5)
+                | is_heavy_hadron(genlep_pa3G, 5)
+                | is_heavy_hadron(genlep_pa4G, 5)
+            )
+
+            Genlep = ak.zip(
+                {
+                    "pT": genlep.pt,
+                    "eta": genlep.eta,
+                    "phi": genlep.phi,
+                    "pdgID": genlep.pdgId,
+                    "status": genlep.status,
+                    "mother": ak.fill_none(
+                        ak.where(isB | isD, 5 * isB + 4 * isD, 10 * istau + 100 * isWZ),
+                        0,
+                    ),
+                }
+            )
+
+            ###################
+            #  Gen particles  #
+            ###################
+            # Gen particles of interest in the BTA ttbar workflow
+            genpart = events.GenPart[
+                (events.GenPart.statusFlags & (1 << 7)) > 0
```

Comment:
Suggest to use coffea features `events.GenPart.hasFlags["isHardProcess"]` for better understanding instead of bitwise info